File: src/Aiko.Core/ShimLaunch.cs
namespace Aiko.Core;

public enum ConfigVariable
{
    /// Leave CLAUDE_CONFIG_DIR as the terminal has it: Aiko knows nothing better.
    Keep,

    /// Set it to the folder of the environment.
    Set,

    /// Remove it, so Claude Code uses .claude the way it does with no variable at all. Setting it
    /// to .claude by hand is not the same: Claude Code then keeps .claude.json in another place.
    Clear,
}

/// What the shim does with one start of Claude Code.
public sealed record ShimPlan(AikoEnvironment? Environment, ConfigVariable Variable, string? ConfigDirectory, bool Explicit)
{
    public static readonly ShimPlan PassThrough = new(null, ConfigVariable.Keep, null, false);
}

/// Decides which environment a start of Claude Code belongs to.
///
/// The shim is one program under several names. Started as "claude", it follows the project
/// folder bindings. Started as a command such as "aiko-work", it runs that environment whatever
/// the folder says: an explicit command wins over a binding (D-159).
public static class ShimLaunch
{
    public const string ClaudeName = "claude";

    public static ShimPlan Decide(
        string invokedAs,
        string workingDirectory,
        EnvironmentSettings settings,
        string userProfile)
    {
        var name = Path.GetFileNameWithoutExtension(invokedAs);

        if (name.Equals(ClaudeName, StringComparison.OrdinalIgnoreCase))
        {
            var bound = ProjectBinding.EnvironmentFor(workingDirectory, settings, userProfile);
            return PlanFor(bound, userProfile, isExplicit: false);
        }

        var chosen = settings.Environments.FirstOrDefault(e => e.Command.Equals(name, StringComparison.OrdinalIgnoreCase));

        // A command left over from a renamed or removed environment. Starting Claude Code as it
        // is beats refusing to start at all.
        return chosen is null ? ShimPlan.PassThrough : PlanFor(chosen, userProfile, isExplicit: true);
    }

    private static ShimPlan PlanFor(AikoEnvironment? environment, string userProfile, bool isExplicit)
    {
        if (environment is null || environment.ConfigDirectories.Count == 0)
        {
            return ShimPlan.PassThrough;
        }

        var folder = environment.ConfigDirectories[0];
        return ClaudeConfigFolder.IsDefault(folder, userProfile)
            ? new ShimPlan(environment, ConfigVariable.Clear, null, isExplicit)
            : new ShimPlan(environment, ConfigVariable.Set, folder, isExplicit);
    }
}

File: mac/Sources/AikoKit/ShimLaunch.swift
import Foundation

public enum ConfigVariable: Sendable, Equatable {
    /// Leave CLAUDE_CONFIG_DIR as the terminal has it: Aiko knows nothing better.
    case keep

    /// Set it to the folder of the environment.
    case set

    /// Remove it, so Claude Code uses .claude the way it does with no variable at all. Setting it
    /// to .claude by hand is not the same: Claude Code then keeps .claude.json in another place.
    case clear
}

/// What the shim does with one start of Claude Code.
public struct ShimPlan: Sendable, Equatable {
    public let environment: AikoEnvironment?
    public let variable: ConfigVariable
    public let configDirectory: String?
    public let isExplicit: Bool

    public init(_ environment: AikoEnvironment?, _ variable: ConfigVariable, _ configDirectory: String?, _ isExplicit: Bool) {
        self.environment = environment
        self.variable = variable
        self.configDirectory = configDirectory
        self.isExplicit = isExplicit
    }

    public static let passThrough = ShimPlan(nil, .keep, nil, false)
}

/// Decides which environment a start of Claude Code belongs to.
///
/// The shim is one program under several names. Started as "claude", it follows the project
/// folder bindings. Started as a command such as "aiko-work", it runs that environment whatever
/// the folder says: an explicit command wins over a binding (D-159).
public enum ShimLaunch {
    public static let claudeName = "claude"

    public static func decide(
        _ platform: PlatformConventions,
        invokedAs: String,
        workingDirectory: String,
        settings: EnvironmentSettings,
        userProfile: String
    ) -> ShimPlan {
        let name = PathText.fileNameWithoutExtension(invokedAs)

        if name.caseInsensitiveCompare(claudeName) == .orderedSame {
            let bound = ProjectBinding.environmentFor(
                platform, workingDirectory: workingDirectory, settings: settings, userProfile: userProfile)
            return planFor(platform, bound, userProfile, isExplicit: false)
        }

        let chosen = settings.environments.first { $0.command.caseInsensitiveCompare(name) == .orderedSame }

        // A command left over from a renamed or removed environment. Starting Claude Code as it
        // is beats refusing to start at all.
        return chosen == nil ? .passThrough : planFor(platform, chosen, userProfile, isExplicit: true)
    }

    private static func planFor(
        _ platform: PlatformConventions,
        _ environment: AikoEnvironment?,
        _ userProfile: String,
        isExplicit: Bool
    ) -> ShimPlan {
        guard let environment, let folder = environment.configDirectories.first else {
            return .passThrough
        }

        return ClaudeConfigFolder.isDefault(platform, folder, userProfile)
            ? ShimPlan(environment, .clear, nil, isExplicit)
            : ShimPlan(environment, .set, folder, isExplicit)
    }
}
